lib/blob_store/src/bitmask/gaps.rs

```diff
@@ -186,72 +186,70 @@ impl BitmaskGaps {
-                .find_map(|(region_id, gap)| {
-                    if gap.max as usize >= num_blocks as usize {
-                        Some(region_id as RegionId..(region_id + 1) as RegionId)
-                    } else {
-                        None
-                    }
-                }),
-            // we need to merge at least 2 regions
-            window_size => self.find_merged_gap(window_size, num_blocks),
-        };
-
-        if fits_in_min_regions.is_some() {
-            return fits_in_min_regions;
+        if self.len() < min_regions_needed {
+            return None;
         }
 
-        // try to find gap by merging one more region (which is the maximum regions we may need for the value)
-        let window_size = regions_needed + 1;
+        let (mut head, mut tail) = (0, 0);
 
-        self.find_merged_gap(window_size, num_blocks)
-    }
+        // Slide head/tail over regions until we find a gap
+        'slider: while head < self.len() {
+            // If we need more/less regions, advance head/tail
+            let region_count = head.saturating_sub(tail) + 1;
+            if region_count < min_regions_needed {
+                head += 1;
+                continue;
+            }
+            if region_count > max_regions_needed {
+                tail += 1;
+                continue;
+            }
+
+            // If enough blocks in just the head, return
+            let head_region = self.get(head).unwrap();
+            if min_regions_needed == 1 && u32::from(head_region.max) >= num_blocks {
+                return Some(head as u32..head as u32 + 1);
+            }
+
+            let mut blocks = u32::from(head_region.leading);
+
+            // If enough blocks including middle regions, return
+            let middle_regions = (tail..=head)
+                .rev()
+                .skip(1)
+                .take(region_count.saturating_sub(2));
+            for region_index in middle_regions {
+                let region = self.get(region_index).unwrap();
+                blocks += u32::from(region.trailing);
 
-    /// Find a gap in the bitmask that is large enough to fit `num_blocks` blocks, in a merged window of regions.
-    fn find_merged_gap(&self, window_size: usize, num_blocks: u32) -> Option<Range<RegionId>> {
-        debug_assert!(window_size >= 2, "window size must be at least 2");
-
-        self.as_slice()
-            .windows(window_size)
-            .enumerate()
-            .find_map(|(start_region_id, gaps)| {
-                // make sure the middle regions are all free
-                let middle_regions = &gaps[1..window_size - 1];
-                if middle_regions
-                    .iter()
-                    .any(|gap| gap.max as usize != self.config.region_size_blocks)
-                {
-                    return None;
+                if blocks >= num_blocks {
+                    return Some(region_index as u32..head as u32 + 1);
                 }
```

Comment:
This check only applies until `(head - region_index) >= min_regions_needed`, but it is fine since we cannot avoid to check the region's `trailing` field, and the check is super cheap.